Rework Books/Magento requests

JonNorman opened the pull request:
## Background
There is a commercial endpoint that takes an ISBN and returns a JSON representation of the associated book. To do this, it uses a third party service called Magento.

Magento is not reliable, it is often very slow, regularly causes our latency alarms to trigger, which in turns blocks builds, which in turn has prompted some to term it the "most annoying thing since, like, forever".

In addition to this there is _a lot_ of code surrounding this service that can probably be culled. The main purpose of this PR is to reduce the latency of requests. Code culling can occur in a separate PR.

## What does this change?
This PR achieves the following:

1) Sets up an actor that can use the Magento Service to look-up book data and then cache it locally using a `Map`.

2) Requests to the commercial endpoint will now ask the actor for the book data. If actor has the book in the cache then it returns it, otherwise it returns `None` and only then uses Magento to find the book data and update its cache.

3) Multiple requests for the same ISBN that are received before the first Magento request has completed will now be ignored.

4) The ISBN will be validated prior to going to Magento to search for it.

## What is the value of this and can you measure success?
Builds won't break and there is less of an overhead when working with Magento.

## Does this affect other platforms - Amp, Apps, etc?
Nope

## Screenshots
Log output when quickly requesting the same ISBN:
```
2017-04-21 23:59:12,794 [application-akka.actor.default-dispatcher-5] INFO  m.merchandise.books.BookJsonActor - Cache miss for ISBN: 9781784630935.
2017-04-21 23:59:12,795 [application-akka.actor.default-dispatcher-5] INFO  m.merchandise.books.BookJsonActor - Looking up ISBN 9781784630935 with Magento Service.
2017-04-21 23:59:12,801 [application-akka.actor.default-dispatcher-5] INFO  c.m.merchandise.books.MagentoService - Looking up book with ISBN 9781784630935 ...
2017-04-21 23:59:13,464 [application-akka.actor.default-dispatcher-2] INFO  m.merchandise.books.BookJsonActor - Cache miss for ISBN: 9781784630935.
2017-04-21 23:59:13,465 [application-akka.actor.default-dispatcher-2] INFO  m.merchandise.books.BookJsonActor - ISBN lookup is already due for 9781784630935. Not running Magento Service.
2017-04-21 23:59:14,613 [application-akka.actor.default-dispatcher-8] INFO  m.merchandise.books.BookJsonActor - Cache hit for ISBN: 9781784630935.
```

The CURL requests (3 in quick succession):
```
curl -v http://localhost:9000/commercial/books/api/book.json?t=9781784630935
*   Trying ::1...
* Connected to localhost (::1) port 9000 (#0)
> GET /commercial/books/api/book.json?t=9781784630935 HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Date: Fri, 21 Apr 2017 13:59:12 GMT
< ETag: W/"hash1"
< Vary: Origin,Accept
< Expires: Fri, 21 Apr 2017 13:59:42 GMT
< Cache-Control: max-age=30, stale-while-revalidate=3, stale-if-error=864000
< Surrogate-Key: 7c46324cc4187d5dea508ecc767be463
< X-Gu-Backend-App: commercial
< Surrogate-Control: max-age=30, stale-while-revalidate=3, stale-if-error=864000
< Content-Length: 0
< Content-Type: application/json
<
* Connection #0 to host localhost left intact
```
```
curl -v http://localhost:9000/commercial/books/api/book.json?t=9781784630935
*   Trying ::1...
* Connected to localhost (::1) port 9000 (#0)
> GET /commercial/books/api/book.json?t=9781784630935 HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Date: Fri, 21 Apr 2017 13:59:13 GMT
< ETag: W/"hash1"
< Vary: Origin,Accept
< Expires: Fri, 21 Apr 2017 13:59:43 GMT
< Cache-Control: max-age=30, stale-while-revalidate=3, stale-if-error=864000
< Surrogate-Key: 7c46324cc4187d5dea508ecc767be463
< X-Gu-Backend-App: commercial
< Surrogate-Control: max-age=30, stale-while-revalidate=3, stale-if-error=864000
< Content-Length: 0
< Content-Type: application/json
<
* Connection #0 to host localhost left intact
```

```
curl -v http://localhost:9000/commercial/books/api/book.json?t=9781784630935
*   Trying ::1...
* Connected to localhost (::1) port 9000 (#0)
> GET /commercial/books/api/book.json?t=9781784630935 HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Fri, 21 Apr 2017 13:59:14 GMT
< ETag: W/"hash1006431888418119884"
< Vary: Origin,Accept
< Expires: Fri, 21 Apr 2017 14:00:14 GMT
< Cache-Control: max-age=60, stale-while-revalidate=6, stale-if-error=864000
< Surrogate-Key: 7c46324cc4187d5dea508ecc767be463
< X-Gu-Backend-App: commercial
< Surrogate-Control: max-age=3600, stale-while-revalidate=360, stale-if-error=864000
< Content-Length: 474
< Content-Type: application/json
<
* Connection #0 to host localhost left intact
{"title":"The Clocks in This House All Tell Different Times","author":"Xan Brooks","isbn":"9781784630935","price":8.99,"offerPrice":6.74,"jacketUrl":"https://d935jy3y59lth.cloudfront.net/media/catalog/product/cache/0/image/17f82f742ffe127f42dca9de82fb58b1/media2/4bb20dbb8
```

## Tested in CODE?
No (but I will before merging)
<!-- AB test? https://git.io/v1V0x -->
<!-- AMP question? https://git.io/v1V0p -->
<!-- Does this PR meet the contributing guidelines? https://git.io/v1VEJ -->

JonNorman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Decouple ISBN lookups from Magento](https://trello.com/c/3orcR9uo/99-decouple-isbn-lookups-from-magento)

TBonnin commented on commercial/app/controllers/BookOffersController.scala:
```diff
@@ -25,38 +24,32 @@ class BookOffersController(bookFinder: BookFinder, bestsellersAgent: Bestsellers
       (specificBooks ++ bestsellersAgent.bestsellersTargetedAt(segment)).distinctBy(_.isbn).take(4)
     }
 
+  private def isValidIsbn(isbn: String): Boolean = true
```
always true?

JonNorman replied:
Gah, will correct!

TBonnin commented on commercial/app/controllers/BookOffersController.scala:
```diff
@@ -25,38 +24,32 @@ class BookOffersController(bookFinder: BookFinder, bestsellersAgent: Bestsellers
-          Cached(componentMaxAge)(jsonFormat.nilResult)
+
+    lazy val failedLookupResponse: Result = Cached(30.seconds)(JsonNotFound())(request)
+    lazy val badRequestResponse: Future[Result] = Future.successful(Cached(1.day)(JsonComponent(JsNull))(request))
```
Both have a similar name while being of different type. Could we make this more explicit?

TBonnin commented on commercial/app/controllers/BookOffersController.scala:
```diff
@@ -25,38 +24,32 @@ class BookOffersController(bookFinder: BookFinder, bestsellersAgent: Bestsellers
-      }
-    } getOrElse {
-      Future.successful(NoCache(jsonFormat.nilResult.result))
+      case Some(invalidIsbn) => log.error(s"Book lookup called with invalid ISBN '$invalidIsbn'. Returning empty response."); badRequestResponse
```
Let's use new lines. It is easier to read imho

JonNorman commented:
Hey @TBonnin! I think I've addressed all of your review comments (thanks for taking a look).

The diff now looks a fair bit larger as it turns out that `BookFinder` returning a `Future[JsValue]` had a knock-on effect to a lot of the commercial Merchandise-handling code. Now that `BookFinder` just returns an `Option[JsValue]` a lot of that could be stripped out.

I also cleaned up a few bits where I could (but left some things for a future PR, such as looking at the `FeedReader` setup which seems to only be supporting the Book component).

TBonnin commented on commercial/app/controllers/BookOffersController.scala:
```diff
@@ -20,46 +17,35 @@ class BookOffersController(bookFinder: BookFinder, bestsellersAgent: Bestsellers
+  private def booksSample(isbns: Seq[String], segment: Segment): Seq[Book] =
+    (bestsellersAgent.getSpecificBooks(isbns) ++ bestsellersAgent.bestsellersTargetedAt(segment)).distinctBy(_.isbn).take(4)
+
+  private def isValidIsbn(isbn: String): Boolean = (isbn forall (_.isDigit)) && (isbn.length == 10 || isbn.length == 13)
```
Ae all those parenthesis necessary? 
Also I would stick with the dot notation to call methods and not mix dot and space in the same statement. ie: `isbn.foreall(isDigit) && (isbn.length == 10 || isbn.length == 13)`

JonNorman replied:
>Are all those parenthesis necessary?
If I'm not using dot notation then I believe so, yes - the `&&` appears to bind more closely than the `forall` when not using dot notation.

> I would stick with the dot notation to call methods and not mix dot and space in the same statement.

I generally agree with that sentiment but in this case I am only using dot notation on instance methods i.e. the `String` instance method of `isDigit`. I do not think that it can be called in the method you suggest above: `isbn.forall(isDigit)` as `isDigit` is an instance method, not a class/object method.

Therefore I think I'm being consistent in using space notation in chaining operations and the `_.` notation when invoking instance methods. I know that `post-fix` operations can be imported and then used but this just seems to muddy the waters in my opinion, hence I've kept it as `isbn.length` and not `isbn length`.

TBonnin replied:
Fine. I personally don't like the space notation but as you said you stay consistent in this file. Codebase-wise is another matter though 😋 )
Anyway keep it this way.

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
+          magentoService.findByIsbn(isbn) map (_.foreach { json: JsValue => cache alter { _ + (isbn -> json) } })
+          lookupJobs alter { _ + isbn }
+        }
+        None
```
Is the logging in this function adding any value?

JonNorman replied:
It helped a lot when trying to see where a component was coming from, however it might add a lot of noise to the PROD logs.

What level of logs does Kibana consume? I could make these `log.debug` statements instead or just remove them. If we have any trouble with this functionality in future then presumably whoever looks at it can add their own temporary logging.

TBonnin replied:
Only `info`, `warn` and `errors` are sent to kibana. 
If those are mostly debugging statement, using `log.debug` makes sense

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+  private lazy val lookupJobs = AkkaAgent(Set.empty[String])
+
+  def get(isbn: String)(implicit magentoService: MagentoService, executionContext: ExecutionContext): Option[JsValue] =
```
We assume with this function that the json for a given book will never change once it has been added to the cache. Is it always true?

JonNorman replied:
Yes I believe it is.

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+  private lazy val lookupJobs = AkkaAgent(Set.empty[String])
```
What is the purpose of storing the lookupJobs?

JonNorman replied:
It is to prevent the Magento Lookup of the same ISBN and is used [here](https://github.com/guardian/frontend/pull/16533/files/482361cfc65400b08035d4016d71c4c3ed740f64#diff-b8caa4a82242d9b3eece8ab4bdf98828R42).

Scenario: we get request `A` in for ISBN `123`. We don't have it in the cache so we use Magento to look it up. Whilst the Magento service is doing this (can take seconds), request `B` comes in from another user for the same ISBN `123`. We needn't go to Magento again as we know that it is already being retrieved.

JonNorman replied:
I've just thought that, if the magento lookup fails, then we may not want to add it to our local cache - I will add a fix to remedy this...

TBonnin replied:
Would it make sense to only store pending lookup? (ie: removing it once magento successful request returns).

Also how much benefit does adding this extra complexity gives us? How often a request would hit this endpoint while a magento request for the same book is pending?

JonNorman replied:
> Would it make sense to only store pending lookup? (ie: removing it once magento successful request returns).

I was thinking of that but it could lead to race conditions since the `cache` and `lookupJobs` are using different `AkkaAgent`s and there isn't a simple way to ensure updates to one will follow updates to the other made later, we could introduce a third agent but I think we might then suffer from death-by-agent.

>Also how much benefit does adding this extra complexity gives us? How often a request would hit this endpoint while a magento request for the same book is pending?

I don't think it will happen that often? Perhaps 3 times per book? Depends on how much traffic our book pages are generating. Happy to strip it out if it is deemed too much complexity for too little reward - was just trying to be a good citizen 😁

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
+
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
```
Just curious: How many book info we expect to store in this cache?

JonNorman replied:
In theory it can be as many as the number of (distinct) book reviews we have [on the site](http://content.guardianapis.com/search?tag=tone%2Freviews%2Cbooks%2Fbooks&api-key=test) (32,630). I think that this is manageable for the little `Map`, but perhaps it would be prudent to add a TTL for each book in the cache and do a bit of pruning when updating the agent. What do you think?

TBonnin replied:
Adding expiration would be safer but I would greatly increase the complexity. 🤔 
How much memory would those 32k books info represent?

JonNorman replied:
Looking at a single example of a `Book` it is 952 Bytes(!) so I think we are probably good on this front for now :)
![image](https://cloud.githubusercontent.com/assets/1821099/25334670/82a8d29a-28e7-11e7-8c70-f51a239943f3.png)

TBonnin replied:
which mean 30ish mb of data in memory in case all books are being cached. We are probably safe here. 
I am sure the number of book reviews increases over time but probably not at a very fast rate

JonNorman commented:
As an update/summary on this:

### Original Problem
When an ISBN is requested - `https://api.nextgen.guardianapps.co.uk/commercial/books/api/book.json?t=0123456789` - if Fastly encounters a cache miss, then the commercial app will be hit, which will then inspect its own application level cache before looking it up in the Magento service.

In practice this results in long latencies as when the application level cache misses, the response depends on the magento service to return a result in a decent time, which is rare.

### Proposed Fix (this PR)
The commercial app should check its (now vastly simplified) local cache to see if the book is in there. If it isn't then _immediately return a `404` and then go to Magento to update the local cache_. This only has to happen once and then Fastly will cache the response.

After discussing with @TBonnin, another potential fix would be to proxy Magento through Fastly, with a short timeout such that, when a request for an ISBN comes through to commercial, it requests from Magento as before but behind Fastly and with a short timeout such that the latency of the entire request chain will be kept low. If Magento is responsive then Fastly should return the book within the timeout otherwise the commercial request will timeout and Fastly will eventually cache the response from Magento.

Having thought about these two solutions a bit this morning, I'm leaning towards the former (local cache). It seems simpler, by reducing the services involved in the entire transaction and, given that we already have the commercial books endpoint behind Fastly, we just need to tighten up the latency on the first few requests for each ISBN. Incidentally using the local cache also allows a fair amount of simplification in other merchandise controllers.

@TBonnin - happy to come over and discuss again in person, but if you don't mind much either way then I can go ahead and merge this, see how it performs and take a view after then.

TBonnin commented:
I am fine with it.
However I find it unnecessary to implement our own caching (with all the related complications: invalidations, memory cost, etc...) when a CDN can do the same job better.

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,64 @@ package commercial.model.merchandise.books
+          }
+        }
+        None
+      }
```
I would chain all those Futures to make sure you avoid race conditions.
Something like this:
```
         val fAddPending = isbnsToBeAddedToCache.alter(_ + isbn)
          val fLookup: Future[Option[JsValue]] = magentoService.findByIsbn(isbn)
          val fetchAndUpdateBook = for {
            _ <- fAddPending
            lookup <- fLookup
            newBookCache <- lookup.map(json => cache alter { _ + (isbn -> json) }) match {
              case Some(f) => f.map(Some(_))
              case None    => Future.successful(None)
            }
          } yield newBookCache

          fetchAndUpdateBook.andThen { case _ =>
            isbnsToBeAddedToCache alter { _ - isbn }
          }
```

JonNorman replied:
Yep I can do that - though [Akka guarantees](http://doc.akka.io/docs/akka/current/scala/agents.html#Updating_Agents__send___alter_) that the order of operations of subsequent calls to `alter` on the same agent will be honured. So I don't think that it is strictly required here.

JonNorman replied:
What race condition do you think could exist as it currently stands?

TBonnin replied:
More than race conditions actually, I am more worried about some of those operations to be independent from each other. 
Do we want to execute magento request if the first `alter` fails? Might not be a big problem though.
We also want to remove the `isbn` from `isbnsToBeAddedToCache` not only when the lookup return `None` but also when it fails, as well as when the `cache` `alter` fails. 
Does that make sense?

JonNorman replied:
Yep that makes sense - thanks for the explanation!

JonNorman commented:
@TBonnin I followed your advice on linking the Futures [here](https://github.com/guardian/frontend/pull/16533/commits/47712d8a3a9f6313f2081ab7a6bc0130122cb79f). Having done that I reassessed the value of keeping track of the jobs that are currently being searched and found that the saving probably isn't worth the added complexity (as you suggested before!), so the [subsequent commit](https://github.com/guardian/frontend/pull/16533/commits/057ae07adc8287adaef50285a0ec38cfa7cd1d20) just removes that functionality, which I think is far easier to comprehend at very little performance cost.

PRBuilds commented:
PR build results:

screenshots
[mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16533/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png) • [wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16533/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16533/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16533/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png)
    
exceptions (0)
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16533/thrown-exceptions.js)    

webpagetest (2)
[performanceComparisonSummary.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16533/performanceComparisonSummary.txt)

-automated message

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,47 @@ package commercial.model.merchandise.books
+            case Success(None) => log.warn(s"Magento unable to find book for $isbn.")
+            case Success(Some(json: JsValue)) => cache alter { _ + (isbn -> json) }
+          }
+          None
```
The first pattern matching is probably not necessary here since you are actually returning the result of `cache.get.get(isbn)` as is.
What about:
```
val bookJson = cache.get.get(isbn)
if(bookJson.isEmpty) {
    magentoService.findByIsbn(isbn) onComplete {
        case Failure(e) => log.error("Magento lookup failed.", e)
        case Success(None) => log.warn(s"Magento unable to find book for $isbn.")
        case Success(Some(json: JsValue)) => cache alter { _ + (isbn -> json) }
    }
}
bookJson
```

TBonnin commented on commercial/app/model/merchandise/books/BookFinder.scala:
```diff
@@ -2,67 +2,47 @@ package commercial.model.merchandise.books
+        case Success(None) => log.warn(s"Magento unable to find book for $isbn.")
+        case Success(Some(json: JsValue)) => cache alter { _ + (isbn -> json) }
+      }
     }
```
`bookJson` needs to be returned here :D

JonNorman replied:
:weary: Sorry about that - lazy on my part. Fixing now.

TBonnin commented:
Thank you for doing this @JonNorman 
Can't wait for not receiving Commercial alerts anymore 😝

JonNorman commented:
No worries at all - thanks for all the review attention - it's a lot better than it was going into this PR 👍

jfsoul commented:
In case you weren't already looking - the early signs are very promising:

![image](https://cloud.githubusercontent.com/assets/8754692/25438879/ac5550b2-2a92-11e7-9702-286cd1d79b30.png)

(orange is latency)

@JonNorman 👏 👏

JonNorman commented:
Ah I missed this but just checked the alarm and it looks pretty good!
![image](https://cloud.githubusercontent.com/assets/1821099/25613087/e4edd1ba-2f24-11e7-8ee3-aea2c81fec2c.png)
